### PR TITLE
8324734: Relax too-strict assert(VM_Version::supports_evex()) in Assembler::locate_operand()

### PR DESCRIPTION
**Pull request by rkennke**
Details see bug report. The gist is that HotSpot downgrades to UseAVX=2 on some processors, and reports supports_evex() == false, but the instruction decoder can still encounter EVEX instructions when (e.g.) hitting a SIGBUS in memset() - which does have EVEX instructions.

Testing:
 - [x] runtime/Unsafe/InternalErrorTest.java
 - [x] tier1

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8324734](https://bugs.openjdk.org/browse/JDK-8324734): Relax too-strict assert(VM_Version::supports_evex()) in Assembler::locate_operand() (**Bug** - P3)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Contributors
 * Vladimir Kozlov `<kvn@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17590/head:pull/17590` \
`$ git checkout pull/17590`

Update a local copy of the PR: \
`$ git checkout pull/17590` \
`$ git pull https://git.openjdk.org/jdk.git pull/17590/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17590`

View PR using the GUI difftool: \
`$ git pr show -t 17590`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17590.diff">https://git.openjdk.org/jdk/pull/17590.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17590#issuecomment-1912244126)

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
The assert is there to make sure HotSpot generates correct code.
What output you get in your test case with this change? You still crash in `memset`.

**Comment by rkennke** [comment]
> The assert is there to make sure HotSpot generates correct code.

I understand, but the decoder here tries to decode non-HotSpot code, which can legitimately use EVEX instructions.

> What output you get in your test case with this change? You still crash in `memset`.

The test-case passes with the change. The test provokes a SIGBUS and checks that it is propely turned into an InternalError. Without the change, it would crash with the assert, with the change the SIGBUS would be handled and turned into the expected InternalError.

What we *could* do is what Jorn suggests in JBS:
"As a long term solution, maybe VM_Version could distinguish between CPU features that are supported by the CPU, and features that are enabled. Then, the decoder could check whether evex is supported, while other code could be changed to check whether use of evex is enabled for hotspot's own code gen."
However, that seems more complex than what I'm currently willing to spend time on. ;-)

**Comment by vnkozlov** [comment]
> VM_Version could distinguish between CPU features that are supported by the CPU
We can start with just EVEX check. It is not big change:
```
$ git diff
diff --git a/src/hotspot/cpu/x86/vm_version_x86.cpp b/src/hotspot/cpu/x86/vm_version_x86.cpp
index df1ea6edd30..8b4ca442b5a 100644
--- a/src/hotspot/cpu/x86/vm_version_x86.cpp
+++ b/src/hotspot/cpu/x86/vm_version_x86.cpp
@@ -809,7 +809,8 @@ void VM_Version::get_processor_features() {
   _stepping = cpu_stepping();
 
   if (cpu_family() > 4) { // it supports CPUID
-    _features = feature_flags();
+    _features = feature_flags(); // It can be changed by VM flags
+    _cpu_features = _features;   // Preserve features
     // Logical processors are only available on P4s and above,
     // and only if hyperthreading is available.
     _logical_processors_per_package = logical_processor_count();
diff --git a/src/hotspot/cpu/x86/vm_version_x86.hpp b/src/hotspot/cpu/x86/vm_version_x86.hpp
index e521a6ee3bc..de86ce51541 100644
--- a/src/hotspot/cpu/x86/vm_version_x86.hpp
+++ b/src/hotspot/cpu/x86/vm_version_x86.hpp
@@ -640,7 +640,7 @@ class VM_Version : public Abstract_VM_Version {
   }
 
   //
-  // Feature identification
+  // Feature identification which can be affected by VM flags
   //
   static bool supports_cpuid()        { return _features  != 0; }
   static bool supports_cmov()         { return (_features & CPU_CMOV) != 0; }
@@ -703,6 +703,11 @@ class VM_Version : public Abstract_VM_Version {
   static bool supports_cet_ss()       { return (_features & CPU_CET_SS) != 0; }
   static bool supports_cet_ibt()      { return (_features & CPU_CET_IBT) != 0; }
 
+  //
+  // Feature identification not affected by VM flags
+  //
+  static bool cpu_supports_evex()         { return (_cpu_features & CPU_AVX512F) != 0; }
+
   // Intel features
   static bool is_intel_family_core() { return is_intel() &&
                                        extended_cpu_family() == CPU_FAMILY_INTEL_CORE; }
diff --git a/src/hotspot/share/runtime/abstract_vm_version.hpp b/src/hotspot/share/runtime/abstract_vm_version.hpp
index d8ffca8de81..05675cc683a 100644
--- a/src/hotspot/share/runtime/abstract_vm_version.hpp
+++ b/src/hotspot/share/runtime/abstract_vm_version.hpp
@@ -54,10 +54,13 @@ class Abstract_VM_Version: AllStatic {
   static const char*  _s_vm_release;
   static const char*  _s_internal_vm_info_string;
 
-  // CPU feature flags.
+  // CPU feature flags which can be restricted by VM flags.
   static uint64_t _features;
   static const char* _features_string;
 
+  // CPU feature flags not affected by VM flags.
+  static uint64_t _cpu_features;
+
   // These are set by machine-dependent initializations
 #ifndef SUPPORTS_NATIVE_CX8
   static bool         _supports_cx8;
```

**Comment by rkennke** [comment]
> > VM_Version could distinguish between CPU features that are supported by the CPU
> > We can start with just EVEX check. It is not big change:
> 
> ```
> $ git diff
> diff --git a/src/hotspot/cpu/x86/vm_version_x86.cpp b/src/hotspot/cpu/x86/vm_version_x86.cpp
> index df1ea6edd30..8b4ca442b5a 100644
> --- a/src/hotspot/cpu/x86/vm_version_x86.cpp
> +++ b/src/hotspot/cpu/x86/vm_version_x86.cpp
> @@ -809,7 +809,8 @@ void VM_Version::get_processor_features() {
>    _stepping = cpu_stepping();
>  
>    if (cpu_family() > 4) { // it supports CPUID
> -    _features = feature_flags();
> +    _features = feature_flags(); // It can be changed by VM flags
> +    _cpu_features = _features;   // Preserve features
>      // Logical processors are only available on P4s and above,
>      // and only if hyperthreading is available.
>      _logical_processors_per_package = logical_processor_count();
> diff --git a/src/hotspot/cpu/x86/vm_version_x86.hpp b/src/hotspot/cpu/x86/vm_version_x86.hpp
> index e521a6ee3bc..de86ce51541 100644
> --- a/src/hotspot/cpu/x86/vm_version_x86.hpp
> +++ b/src/hotspot/cpu/x86/vm_version_x86.hpp
> @@ -640,7 +640,7 @@ class VM_Version : public Abstract_VM_Version {
>    }
>  
>    //
> -  // Feature identification
> +  // Feature identification which can be affected by VM flags
>    //
>    static bool supports_cpuid()        { return _features  != 0; }
>    static bool supports_cmov()         { return (_features & CPU_CMOV) != 0; }
> @@ -703,6 +703,11 @@ class VM_Version : public Abstract_VM_Version {
>    static bool supports_cet_ss()       { return (_features & CPU_CET_SS) != 0; }
>    static bool supports_cet_ibt()      { return (_features & CPU_CET_IBT) != 0; }
>  
> +  //
> +  // Feature identification not affected by VM flags
> +  //
> +  static bool cpu_supports_evex()         { return (_cpu_features & CPU_AVX512F) != 0; }
> +
>    // Intel features
>    static bool is_intel_family_core() { return is_intel() &&
>                                         extended_cpu_family() == CPU_FAMILY_INTEL_CORE; }
> diff --git a/src/hotspot/share/runtime/abstract_vm_version.hpp b/src/hotspot/share/runtime/abstract_vm_version.hpp
> index d8ffca8de81..05675cc683a 100644
> --- a/src/hotspot/share/runtime/abstract_vm_version.hpp
> +++ b/src/hotspot/share/runtime/abstract_vm_version.hpp
> @@ -54,10 +54,13 @@ class Abstract_VM_Version: AllStatic {
>    static const char*  _s_vm_release;
>    static const char*  _s_internal_vm_info_string;
>  
> -  // CPU feature flags.
> +  // CPU feature flags which can be restricted by VM flags.
>    static uint64_t _features;
>    static const char* _features_string;
>  
> +  // CPU feature flags not affected by VM flags.
> +  static uint64_t _cpu_features;
> +
>    // These are set by machine-dependent initializations
>  #ifndef SUPPORTS_NATIVE_CX8
>    static bool         _supports_cx8;
> ```

Nice, thank you! That works (with some minor modification). I've changed the PR from removing the assert to relaxing it to CPU feature check.

**Review by vnkozlov — COMMENTED** [review]
Good. I will submit our testing before approval.
And this needs second review.

File: src/hotspot/cpu/x86/vm_version_x86.hpp
Comment (shipilev) [inline src/hotspot/cpu/x86/vm_version_x86.hpp:709]: Indenting is a bit off here. I think the opening brace should be aligned with the others on the top.

File: src/hotspot/cpu/x86/vm_version_x86.cpp
Comment (shipilev) [inline src/hotspot/cpu/x86/vm_version_x86.cpp:812]:
```suggestion
    _features = feature_flags(); // These can be changed by VM settings
```

File: src/hotspot/cpu/x86/vm_version_x86.hpp
Comment (shipilev) [inline src/hotspot/cpu/x86/vm_version_x86.hpp:643]:
```suggestion
  // Feature identification which can be affected by VM settings
```

File: src/hotspot/share/runtime/abstract_vm_version.hpp
Comment (shipilev) [inline src/hotspot/share/runtime/abstract_vm_version.hpp:57]:
```suggestion
  // CPU feature flags, can be affected by VM settings.
```

File: src/hotspot/share/runtime/abstract_vm_version.hpp
Comment (shipilev) [inline src/hotspot/share/runtime/abstract_vm_version.hpp:61]:
```suggestion
  // Original CPU feature flags, not affected by VM settings.
```

**Review by shipilev — APPROVED** [review]
This looks fine, with nits.

/reviewers 2

**Comment by rkennke** [comment]
> This looks fine, with nits.

Thanks for the review! I fixed all the mentioned issues.

**Review by vnkozlov — APPROVED** [review]
My testing passed.

**Review by shipilev — APPROVED** [review]
Looks fine to me.

**Comment by rkennke** [comment]
Thanks, Vladimir and Aleksey!

/intergrate

**Pull request closed without merging**
